presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java

```diff
@@ -121,55 +117,144 @@ public void testChecksumQuery()
+                ImmutableMap.<String, Object>builder()
+                        .put("bigint_checksum", new SqlVarbinary(new byte[] {0x1a}))
+                        .put("varchar_checksum", new SqlVarbinary(new byte[] {0x1b}))
+                        .build());
+        assertEquals(
+                checksumValidator.getMismatchedColumns(columns, controlChecksum, testChecksum),
+                ImmutableMap.builder()
+                        .put(BIGINT_COLUMN, new ColumnMatchResult(false, "control(checksum: 0a) test(checksum: 1a)"))
+                        .put(VARCHAR_COLUMN, new ColumnMatchResult(false, "control(checksum: 0b) test(checksum: 1b)"))
+                        .build());
     }
 
     @Test
-    public void testColumnsMismatched()
+    public void testFloatingPoint()
     {
+        List<Column> columns = ImmutableList.of(DOUBLE_COLUMN, REAL_COLUMN);
+        ChecksumResult controlChecksum = new ChecksumResult(
+                5,
+                ImmutableMap.<String, Object>builder()
+                        .putAll(FLOATING_POINT_COUNTS)
+                        .put("double_sum", 1.0)
+                        .put("real_sum", 1.0)
+                        .build());
+
+        // Matched
         ChecksumResult testChecksum = new ChecksumResult(
                 5,
                 ImmutableMap.<String, Object>builder()
-                        .put("bigint_checksum", new SqlVarbinary(new byte[] {0x1a}))
-                        .put("varchar_checksum", new SqlVarbinary(new byte[] {0x1b}))
+                        .putAll(FLOATING_POINT_COUNTS)
+                        .put("double_sum", 1 + RELATIVE_ERROR_MARGIN)
+                        .put("real_sum", 1 - RELATIVE_ERROR_MARGIN + RELATIVE_ERROR_MARGIN * RELATIVE_ERROR_MARGIN)
+                        .build());
+        assertTrue(checksumValidator.getMismatchedColumns(columns, controlChecksum, testChecksum).isEmpty());
+
+        // Mismatched
+        testChecksum = new ChecksumResult(
+                5,
+                ImmutableMap.<String, Object>builder()
                         .put("double_sum", 1.0)
                         .put("double_nan_count", 0L)
-                        .put("double_pos_inf_count", 53L)
-                        .put("double_neg_inf_count", 54L)
+                        .put("double_pos_inf_count", 3L)
+                        .put("double_neg_inf_count", 4L)
+                        .put("real_sum", 1.0)
+                        .put("real_nan_count", 2L)
+                        .put("real_pos_inf_count", 0L)
+                        .put("real_neg_inf_count", 4L)
+                        .build());
+        assertEquals(
+                checksumValidator.getMismatchedColumns(columns, controlChecksum, testChecksum),
+                ImmutableMap.builder()
+                        .put(DOUBLE_COLUMN, new ColumnMatchResult(false, "control(NaN: 2, +infinity: 3, -infinity: 4) test(NaN: 0, +infinity: 3, -infinity: 4)"))
```

Comment:
Wondering if we should make this string static constant and put together with `FLOATING_POINT_COUNTS`, so if we need to modify them later it would be easier.

Reply:
The Strings are slightly different in each case, so I think it is more readable to inline.